### PR TITLE
Added yield-effects rule

### PR DESCRIPTION
**Pull request by dejesus2010**
This rule checks to ensure all effects imported are yielded.

It is ported over from the untyped project [eslint-plugin-redux-saga](https://github.com/pke/eslint-plugin-redux-saga).

In this commit I added mocha tests for the new rule and moved existing files
under a `/lib/rules` directory as an attempt to tidy up the project directory a bit.

### CODE REVIEW

**Comment by dejesus2010** [comment]
Addresses issue #5

**Comment by danielweck** [comment]
great, thank you :)

File: index.js [hunk 1]
```diff
@@ -1,8 +1,9 @@
-'use strict';
+"use strict";
```
Comment (SantoJambit) [inline index.js]: a bit weird to just completely change a projects code-style in a PR that has nothing to do with code-style. How about adding prettier or something to enforce the style at least?

File: lib/rules/yield-effects.js [hunk 2]
```diff
@@ -0,0 +1,85 @@
+"use strict";
+
+var Effects = require("../utils/effects");
```
Comment (SantoJambit) [inline lib/rules/yield-effects.js]: please use const or let instead of var (everywhere)

File: lib/rules/yield-effects.js [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+
+var knownEffects = Effects.knownEffects();
+
+module.exports = {
+  meta: {
+    docs: {
+      description: "enforce yield in front of effects",
```
Comment (SantoJambit) [inline lib/rules/yield-effects.js]: yield *

File: lib/rules/yield-effects.js [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+      category: "Possible Errors",
+      recommended: true,
+    },
+    hasSuggestions: true,
```
Comment (SantoJambit) [inline lib/rules/yield-effects.js:15]: the meta object differs a bit from the other rules in terms of attributes. Is one style preferable over the other?

File: lib/rules/yield-effects.js [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+    schema: [],
+  },
+
+  create: function (context) {
```
Comment (SantoJambit) [inline lib/rules/yield-effects.js]: `create(context) {` can be used instead.

File: lib/rules/yield-effects.js [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+      }
+    }
+    function exitFunction(node) {
+      if (node.generator) {
+        --inGeneratorDepth;
+      }
+    }
+    return {
+      ImportDeclaration: function (node) {
+        if (Effects.isEffectImport(node.source.value)) {
```
Comment (SantoJambit) [inline lib/rules/yield-effects.js]: The logic here seems quite similar to the one in delegate-effects. Can they use the same approach or can the code even be extracted into utils?

File: lib/utils/effects.js [hunk 7]
```diff
@@ -0,0 +1,14 @@
+"use strict";
+
+function knownEffects() {
+  return Object.keys(require("typed-redux-saga"));
+}
+
+function isEffectImport(value) {
+  return /^typed-redux-saga?/.test(value);
```
Comment (SantoJambit) [inline lib/utils/effects.js]: Why is "typed-redux-sag" (without the final "a") a valid value here?

File: lib/utils/effects.js [hunk 8]
```diff
@@ -0,0 +1,14 @@
+"use strict";
+
+function knownEffects() {
+  return Object.keys(require("typed-redux-saga"));
```
Comment (SantoJambit) [inline lib/utils/effects.js]: why not export knowEffects as a map instead of as a function?

File: lib/utils/effects.js [hunk 9]
```diff
@@ -0,0 +1,14 @@
...
+}
+
+function isEffectImport(value) {
+  return /^typed-redux-saga?/.test(value);
+}
+
+module.exports = {
+  isEffectImport: isEffectImport,
+  knownEffects: knownEffects,
```
Comment (SantoJambit) [inline lib/utils/effects.js:72]: Both of these could be used in delegate-effects.js as well

File: test/rules/yield-effects-spec.js [hunk 10]
```diff
@@ -0,0 +1,219 @@
...
+          "static* someSaga() {" +
+          "  call(() => {})" +
+          "}" +
+          "}",
+        errors: [
+          {
+            message: "call effect must be yielded",
+            suggestions: [
+              {
+                desc: "Add yield",
```
Comment (SantoJambit) [inline test/rules/yield-effects-spec.js]: yield*

File: test/rules/yield-effects-spec.js [hunk 11]
```diff
@@ -0,0 +1,219 @@
...
+      {
+        code:
+          `import { takeEvery } from 'typed-redux-saga${macro}';\n` +
+          "export const fooSagas = [" +
+          "takeEvery('FOO_A', fooASaga)," +
+          "takeEvery('FOO_B', fooBSaga)];",
+      },
+      {
+        code:
+          `import { call } from 'typed-redux-saga${macro}';\n` +
```
Comment (SantoJambit) [inline test/rules/yield-effects-spec.js]: template literals can span multiple lines:
```
const foobar = `
foo
bar`;
```

I think that would be cleaner for these code snippets.

**Review by SantoJambit — CHANGES_REQUESTED** [review]
Thanks for the work! I have a few remarks

File: lib/utils/effects.js (hunk 7 above)
Comment (dejesus2010) [reply lib/utils/effects.js]: Oh, good catch!

File: index.js (hunk 1 above)
Comment (dejesus2010) [reply index.js]: Yea my bad... I forgot to turn off auto format in my editor and shouldn't have committed this change

File: lib/rules/yield-effects.js (hunk 4 above)
Comment (dejesus2010) [reply lib/rules/yield-effects.js:15]: I don't think one is preferable. I'll update this to be closer to the others while following w/e is suggested on the eslint docs

https://eslint.org/docs/developer-guide/working-with-rules#rule-basics

**Comment by dejesus2010** [comment]
Thanks for the reply @SantoJambit. I have updated the commit with your suggestions

**Comment by dejesus2010** [comment]
So while updating `delegate-effects.js` to use the shared tracker I created in the recent change, I realized `delegate-effects.js` does what this PR is doing... That is: if we're calling a typed effect, ensure that `yield*` is inserted before the effect if no yield exists. I missed that when I was reading through the project.

I added tests to test `delegate-effects.js` and the tests fail when we call an effect without a `yield/yield*`, which is correct. But in my project eslint isn't picking it up... It only picks up the replacement of `yield` to `yield*`.

Any ideas what could be the issue here?

Here's my .eslintrc.json:


```
{
  "extends": ["next/core-web-vitals", "prettier"],
  "plugins": ["@jambit/typed-redux-saga"],
  "overrides": [
    {
      "files": ["./**/*.ts"],
      "excludedFiles": ["./**/*.spec.ts"],
      "rules": {
        "@jambit/typed-redux-saga/use-typed-effects": ["error", "macro"],
        "@jambit/typed-redux-saga/delegate-effects": "error"
      }
    }
  ]
}
```


I'm using the `'typed-redux-saga/macro'` import

**Comment by dejesus2010** [comment]
I submitted a commit that includes the tests I added. I'll wait for your response to decide how to proceed.

**Comment by SantoJambit** [comment]
Thanks

**Pull request merged**